### ISSUE TITLE
Percolation with Bounding Box GeoLocation throws NullPointerException in Lucene

### ISSUE THREAD

**Issue by amscotti** [elastic/elasticsearch#16832]
**Elasticsearch version**: 2.2.0

**JVM version**: 1.8.0_72

**OS version**: Linux 851a91a0d5ad 4.0.7-boot2docker #1 SMP Wed Jul 15 00:01:41 UTC 2015 x86_64 GNU/Linux

Hi, we are currently running a cluster of 2.1.2 and looking to upgrade to 2.2.0, but we are having some issues when it comes to testing our code around percolators. With 2.1.2 everything works fine, but when testing against 2.2.0 we get a return of internal server error with the reason being a null pointer exception. I have tested this with two clean Docker images from the official Elasticsearch images to ensure there's no configuration issues.

The only thing I could find about this issue is an unanswered posting on Stack Overflow here, http://stackoverflow.com/questions/35451052/elastic-search-percolation-with-bounding-box-geolocation-throws-nullpointerexcep

**Steps to reproduce**:
 1 Create Percolator Index

``` bash
curl -XPOST http://192.168.99.101:32770/test -d '{"mappings": {".percolator": {"dynamic": true,"properties": {"id": {"type" : "integer"}}},"test": {"properties": {"location": {"type": "geo_point","lat_lon": true }}}}}'
```

 2 Create Percolator Query

``` bash
curl -XPUT http://192.168.99.101:32770/test/.percolator/alert-1 -d '{"query":{"filtered":{"filter":{"bool":{"must":[{"geo_bounding_box":{"location":{"top_left":[-71.09,42.36],"bottom_right":[-71.085,42.355]},"type":"indexed"}}]}}}}}'
```

 3 Percolator a Document

``` bash
curl -XGET http://192.168.99.101:32772/test/test/_percolate?percolate_format=ids&percolate_index=test -d '{"doc":{"location":{"lon":-71.0875,"lat":42.3575}}}'
```

**Output from 2.1.2**

``` json
{"took":3,"_shards":{"total":5,"successful":5,"failed":0},"total":1,"matches":["alert-1"]}
```

**Output from 2.2.0**

``` json
{"took":6,"_shards":{"total":5,"successful":4,"failed":1,"failures":[{"shard":3,"index":"test","status":"INTERNAL_SERVER_ERROR","reason":{"type":"null_pointer_exception","reason":null}}]},"total":0,"matches":[]}
```

**Provide logs (if relevant)**:

```
RemoteTransportException[[Anelle][127.0.0.1:9301][indices:data/read/percolate[s]]]; nested: PercolateException[failed to percolate]; nested: PercolateException[failed to execute]; nested: NullPointerException;
Caused by: PercolateException[failed to percolate]; nested: PercolateException[failed to execute]; nested: NullPointerException;
        at org.elasticsearch.action.percolate.TransportPercolateAction.shardOperation(TransportPercolateAction.java:180)
        at org.elasticsearch.action.percolate.TransportPercolateAction.shardOperation(TransportPercolateAction.java:55)
        at org.elasticsearch.action.support.broadcast.TransportBroadcastAction$ShardTransportHandler.messageReceived(TransportBroadcastAction.java:268)
        at org.elasticsearch.action.support.broadcast.TransportBroadcastAction$ShardTransportHandler.messageReceived(TransportBroadcastAction.java:264)
        at org.elasticsearch.transport.TransportService$4.doRun(TransportService.java:350)
        at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
Caused by: PercolateException[failed to execute]; nested: NullPointerException;
        at org.elasticsearch.percolator.PercolatorService$4.doPercolate(PercolatorService.java:583)
        at org.elasticsearch.percolator.PercolatorService.percolate(PercolatorService.java:254)
        at org.elasticsearch.action.percolate.TransportPercolateAction.shardOperation(TransportPercolateAction.java:177)
        ... 8 more
Caused by: java.lang.NullPointerException
        at org.apache.lucene.search.GeoPointTermQueryConstantScoreWrapper$1.getDocIDs(GeoPointTermQueryConstantScoreWrapper.java:86)
        at org.apache.lucene.search.GeoPointTermQueryConstantScoreWrapper$1.scorer(GeoPointTermQueryConstantScoreWrapper.java:126)
        at org.apache.lucene.search.LRUQueryCache$CachingWrapperWeight.scorer(LRUQueryCache.java:628)
        at org.apache.lucene.search.BooleanWeight.scorer(BooleanWeight.java:280)
        at org.apache.lucene.search.LRUQueryCache$CachingWrapperWeight.scorer(LRUQueryCache.java:628)
        at org.apache.lucene.search.BooleanWeight.scorer(BooleanWeight.java:280)
        at org.apache.lucene.search.LRUQueryCache$CachingWrapperWeight.scorer(LRUQueryCache.java:628)
        at org.apache.lucene.search.BooleanWeight.scorer(BooleanWeight.java:280)
        at org.apache.lucene.search.LRUQueryCache$CachingWrapperWeight.scorer(LRUQueryCache.java:628)
        at org.elasticsearch.common.lucene.Lucene.exists(Lucene.java:248)
        at org.elasticsearch.percolator.PercolatorService$4.doPercolate(PercolatorService.java:571)
        ... 10 more
```

**Comment by nknize** [comment]
Thanks for the report @amscotti. A couple follow up questions:

1. did you reindex your data?
2. can you post your `geo_point` mapping?

**Comment by amscotti** [comment]
@nknize Thanks for the reply! This is a fully new Elasticsearch setup using Docker to isolate this issue. Following the steps to reproduce we'll create a new index and attempt to percolate a document. Using 2.1.2, the percolation will succeed where in version 2.2.0 you will get a NullPointerException error from the server. Also, the issue isn't just in the Docker image, it also happens on my MacBook running OS X 10.11.3. So I don't think the issue is related to the set up.

The mapping is included in the steps to reproduce as part of the curl command, but here it is pulled out.

```json
{
   "mappings":{
      ".percolator":{
         "dynamic":true,
         "properties":{
            "id":{
               "type":"integer"
            }
         }
      },
      "test":{
         "properties":{
            "location":{
               "type":"geo_point",
               "lat_lon":true
            }
         }
      }
   }
}
```

**Comment by Schartoym** [comment]
Confirm same issue on Windows
**Elasticsearch version**: 2.2.0
**OS version**: Windows Server 2012
Mappings:
```javascript
{
   "mappings":{
      ".percolator":{
         "properties": {
          "query": {
            "type": "object",
            "enabled": false
          }
        }
      },
      "test":{
         "properties":{
            "location":{
               "type":"geo_point",
            }
         }
      }
   }
}
```
Percolation Query:
```javascript
{
  "_index": "test",
  "_type": ".percolator",
  "_id": "550858",
  "_version": 1,
  "found": true,
  "_source": {
    "query": {
      "bool": {
        "must": [          
          {
            "geo_polygon": {
              "location": {
                "points": [
                  {
                    "lat": 55.8118391906254,
                    "lon": 37.526867226289
                  },
                  {
                    "lat": 55.8081657829831,
                    "lon": 37.5251506260451
                  },
                  {
                    "lat": 55.8081326519671,
                    "lon": 37.5251464834297
                  },
                  {
                    "lat": 55.8023318307023,
                    "lon": 37.5263481363022
                  },
                  {
                    "lat": 55.8022898138136,
                    "lon": 37.5263769973219
                  },
                  {
                    "lat": 55.7981320290787,
                    "lon": 37.5316985066514
                  },
                  {
                    "lat": 55.7981018643375,
                    "lon": 37.5317651671351
                  },
                  {
                    "lat": 55.7929765434144,
                    "lon": 37.5530511410252
                  },
                  {
                    "lat": 55.792969394719,
                    "lon": 37.5531051836136
                  },
                  {
                    "lat": 55.7924858389984,
                    "lon": 37.568554672078
                  },
                  {
                    "lat": 55.792516839275,
                    "lon": 37.5686842702771
                  },
                  {
                    "lat": 55.7982298947502,
                    "lon": 37.5774733801441
                  },
                  {
                    "lat": 55.7968776363261,
                    "lon": 37.5790737397055
                  },
                  {
                    "lat": 55.7969071958654,
                    "lon": 37.5793619663257
                  },
                  {
                    "lat": 55.7994213421139,
                    "lon": 37.5803919440669
                  },
                  {
                    "lat": 55.7994746916239,
                    "lon": 37.580384677184
                  },
                  {
                    "lat": 55.8009250884466,
                    "lon": 37.5793547135494
                  },
                  {
                    "lat": 55.8009637661386,
                    "lon": 37.5793021078072
                  },
                  {
                    "lat": 55.8080215965125,
                    "lon": 37.5624792790697
                  },
                  {
                    "lat": 55.8080344294469,
                    "lon": 37.5624365225948
                  },
                  {
                    "lat": 55.8138344258045,
                    "lon": 37.5329107092988
                  },
                  {
                    "lat": 55.8138269869868,
                    "lon": 37.5327768801379
                  },
                  {
                    "lat": 55.8118937516748,
                    "lon": 37.5269403857461
                  },
                  {
                    "lat": 55.8118391906254,
                    "lon": 37.526867226289
                  }
                ]
              }
            }
          }
        ]
      }
    }
  }
}
```
Document percolation
```javascript
GET test/test/_percolate
{
  "doc":{
    "location":{
      "lon": -71.0875,
      "lat": 42.3575
    }
  }
}
```
I've tested same query on Elastic 2.1.2 and it's works correctly

**Comment by nknize** [comment]
So the problem is Lucene's `MemoryIndex` does not support DocValues, and the new GeoPointField requires DocValues for post filtering.  @martijnvg will open a separate issue to add DocValue support and we will reference it on this issue.

**Comment by martijnvg** [comment]
This is the issue that adds doc values support to the MemoryIndex: https://issues.apache.org/jira/browse/LUCENE-7091

**Comment by martijnvg** [comment]
Doc values support to the MemoryIndex has been added and will be included in the Lucene 6.0 release (which hasn't been released yet). Elasticsearch 5.0 will depend on this, so from 5.0 this issue is fixed.

I opened #17105 to backport / fork the fix in 2.x, but there are concerns with forking Memory index and we should not do that.

Luckily there is a workaround for this issue which can help anyone running into this issue. This problems mentioned in this issue only happen if the new geo query implementing are used. We still use the old geo query implementations for indices created before 2.2.0. There is an index setting (`index.version.created`) that we can use to let ES think an index was created on a cluster that ran with an older ES version. If we set it to a version < 2.2.0 then ES uses the older geo query implementations (only for the index we set this created version for).

Note1: You shouldn't set this setting on already created indices before the upgrade to > 2.2. Only for new percolator indices (after to upgrade to > 2.2).
Note2: Only set this setting on indices that hold your percolator queries. Using this setting on all indices, would mean you can't benefit from geo improvements when running regular searches. If your percolator queries co-exist in the same index holding your data then you should move your percolator queries into a dedicated index.

I this case we can set `index.version.created` to `2010299` (version id of version 2.1.2) and `geo_bounding_box`, `geo_polygon` and other geo queries will work as they did before ES 2.2 in the percolator. Example based on the initial post:

```
curl -XPUT "http://localhost:9200/test" -d'
{
  "settings": {
    "index.version.created" : 2010299
  }, 
  "mappings": {
    ".percolator": {
      "dynamic": true,
      "properties": {
        "id": {
          "type": "integer"
        }
      }
    },
    "test": {
      "properties": {
        "location": {
          "type": "geo_point",
          "lat_lon": true
        }
      }
    }
  }
}'
```

**Comment by martijnvg** [comment]
Closing this issue as it has been fixed from 5.0 and beyond.
